File: packages/shared/src/worker-client.ts
/**
 * Cloudflare Worker client for site discovery.
 */

import { z } from 'zod';

/**
 * Worker response schema.
 */
export const WorkerResponseSchema = z.object({
  source: z.string(),
  links: z.array(z.string()).optional(),
  feeds: z.array(z.string()).optional(),
  count: z.number(),
  diagnostics: z.record(z.any()).optional(),
});

export type WorkerResponse = z.infer<typeof WorkerResponseSchema>;

/**
 * Discover request schema.
 */
export const DiscoverRequestSchema = z.object({
  url: z.string().url(),
});

export type DiscoverRequest = z.infer<typeof DiscoverRequestSchema>;

/**
 * RCMP FSJ request schema.
 */
export const RCMPFSJRequestSchema = z.object({
  monthsBack: z.number().optional(),
});

export type RCMPFSJRequest = z.infer<typeof RCMPFSJRequestSchema>;

/**
 * Worker client error.
 */
export class WorkerClientError extends Error {
  constructor(
    message: string,
    public statusCode?: number,
    public response?: unknown
  ) {
    super(message);
    this.name = 'WorkerClientError';
  }
}

/**
 * Worker client options.
 */
export interface WorkerClientOptions {
  baseUrl: string;
  timeout?: number;
}

/**
 * Cloudflare Worker client.
 */
export class WorkerClient {
  private baseUrl: string;
  private timeout: number;

  constructor(options: WorkerClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/$/, ''); // Remove trailing slash
    this.timeout = options.timeout || 30000; // 30s default
  }

  /**
   * Discover new posts on a website.
   */
  async discover(url: string): Promise<WorkerResponse> {
    // Validate input
    const request = DiscoverRequestSchema.parse({ url });

    const response = await this.post('/discover', { url: request.url });
    return WorkerResponseSchema.parse(response);
  }

  /**
   * Get RCMP FSJ posts.
   */
  async rcmpFSJ(monthsBack?: number): Promise<WorkerResponse> {
    // Validate input
    const request = RCMPFSJRequestSchema.parse({ monthsBack });

    const response = await this.post('/profiles/rcmp-fsj', {
      monthsBack: request.monthsBack,
    });
    return WorkerResponseSchema.parse(response);
  }

  /**
   * Make a POST request to the Worker.
   */
  private async post(path: string, body?: unknown): Promise<unknown> {
    const url = `${this.baseUrl}${path}`;

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: body ? JSON.stringify(body) : undefined,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (!response.ok) {
        const text = await response.text();
        throw new WorkerClientError(
          `Worker request failed: ${response.statusText}`,
          response.status,
          text
        );
      }

      return await response.json();
    } catch (error) {
      clearTimeout(timeoutId);

      if (error instanceof WorkerClientError) {
        throw error;
      }

      if ((error as Error).name === 'AbortError') {
        throw new WorkerClientError('Worker request timed out');
      }

      throw new WorkerClientError(
        `Worker request failed: ${(error as Error).message}`
      );
    }
  }
}

/**
 * Create a Worker client instance.
 */
export function createWorkerClient(baseUrl: string): WorkerClient {
  return new WorkerClient({ baseUrl });
}

File: apps/api/app/services/worker_client.py
"""Cloudflare Worker client for site discovery."""

from typing import Any, Optional

import httpx
from pydantic import BaseModel, Field

from app.config import settings


class WorkerResponse(BaseModel):
    """Worker response model."""

    source: str
    links: Optional[list[str]] = None
    feeds: Optional[list[str]] = None
    count: int
    diagnostics: Optional[dict[str, Any]] = None


class WorkerClientError(Exception):
    """Worker client error."""

    def __init__(self, message: str, status_code: Optional[int] = None, response: Any = None):
        super().__init__(message)
        self.status_code = status_code
        self.response = response


class WorkerClient:
    """Cloudflare Worker client."""

    def __init__(self, base_url: str, timeout: int = 30):
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.client = httpx.Client(timeout=timeout)

    def discover(self, url: str) -> WorkerResponse:
        """Discover new posts on a website."""
        response = self._post("/discover", params={"url": url})
        return WorkerResponse(**response)

    def rcmp_fsj(self, months_back: Optional[int] = None) -> WorkerResponse:
        """Get RCMP FSJ posts."""
        params = {"monthsBack": months_back} if months_back is not None else {}
        response = self._post("/profiles/rcmp-fsj", params=params)
        return WorkerResponse(**response)

    def _post(
        self, 
        path: str, 
        body: Optional[dict[str, Any]] = None,
        params: Optional[dict[str, Any]] = None
    ) -> dict[str, Any]:
        """Make a POST request to the Worker."""
        url = f"{self.base_url}{path}"

        try:
            response = self.client.post(url, json=body, params=params)
            response.raise_for_status()
            return response.json()
        except httpx.HTTPStatusError as e:
            raise WorkerClientError(
                f"Worker request failed: {e.response.status_code}",
                status_code=e.response.status_code,
                response=e.response.text,
            )
        except httpx.TimeoutException:
            raise WorkerClientError("Worker request timed out")
        except Exception as e:
            raise WorkerClientError(f"Worker request failed: {str(e)}")

    def close(self) -> None:
        """Close the client."""
        self.client.close()

    def __enter__(self) -> "WorkerClient":
        return self

    def __exit__(self, *args: Any) -> None:
        self.close()


def get_worker_client() -> WorkerClient:
    """Get a Worker client instance."""
    return WorkerClient(settings.worker_base_url)

